userland/examples/tests/acifc/main.c

```diff
@@ -0,0 +1,80 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <analog_comparator.h>
+#include <timer.h>
+#include <tock.h>
+
+static void comparison_polling(uint8_t ac) {
+  uint count = 0;
+  if (ac > 3) {
+    printf("Please choose either ac 0 or 1 for hail, and 2 or 3 for imix.");
+  }else {
+    while (1) {
+      count++;
+      bool result = comparison(ac);
+      printf("Try %d. Result = %d.\n", count, result);
+      if (result == 1) {
+        printf("This implies Vinp > Vinn!\n\n");
+      }else {
+        printf("This implies Vinp < Vinn!\n\n");
+      }
+      delay_ms(1000);
+    }
+  }
+}
+
+static void window_comparison_polling(uint8_t window) {
+  uint count = 0;
+  if (window > 1) {
+    printf("Please choose window 0 for hail and 1 for imix.");
+  }else {
+    while (1) {
+      count++;
+      bool result = window_comparison(window);
+      printf("Try %d. Result = %d.\n", count, result);
+      if (result == 1) {
+        printf("This implies Vacbn_x+1 < Vcommon < Vacap_x!\n\n");
+      }else {
+        printf("This implies Vcommon < Vacan_x+1 or Vcommon > Vacap_x\n\n");
+      }
+      delay_ms(1000);
+    }
+  }
+}
+
+int main(void) {
+  printf("\nACIFC test application\n");
+
+  if (!acifc_exists()) {
+    printf("ACIFC driver does not exist\n");
+    exit(1);
+  }
+
+  // Set mode according to which implementation you want
+  uint8_t mode = 0;
+
+  // Choose your comparator. AC = 0 corresponds to PA06 and PA07, whereas ac = 1
+  // corresponds to PB02 and PB03. On the hail these are the pins DAC and WKP,
+  // and AC2 and AC3 respectively.
+  uint8_t ac = 1;
```

Comment:
Actually, I don't really see why this isn't general enough. The ADC has this channel-structure because it loops through all of the channels sequentially in userland, whereas for an Analog Comparator you generally only need to use one or two at the same time (which most chips have for sure). Right now, this "index" is the number you select in userland. What would be the merit of adding another structure in the capsule to translate to a channel type? That the code would throw an out of bounds error when someone tries to access an index outside of the accepted range?

Reply:
So the channel structure isn't just to allow checking on the selections from userland. The additional feature they provide is limiting capsules to only using channel selections given to them from a board `main.rs` file.

For example, a hypothetical battery-sensing capsule would want to be hooked up to the battery voltage and compare that against some reference voltage. The pins/channels to make that configuration work have to be set by the board `main.rs` file since they will be specific to each design. So from the battery-sensing capsule's perspective, it's given an opaque handle that signifies "the comparator configuration that senses the battery". Without being given this handle, it's never going to be possible to interchangeably function on different platforms.

From this perspective, it's irrelevant whether that handle is a number or a type on the inside. What is important though, is that the handle be some object which can only be created by the board file and only be consumed by the chip driver. Using a `usize` as the structure would hypothetically allow the capsule to craft its own handles, which it is not in a position to do (since the capsule can't know the configuration for every board).